### PR TITLE
parser: Support ADDDATE and SUBDATE function

### PR DESCRIPTION
**Pull request by zimulala**
#419

### CODE REVIEW

**Comment by siddontang** [comment]
LGTM

**Comment by qiuyesuifeng** [comment]
From this refer doc: `https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_adddate`
u can get that
```
ADDDATE(date,INTERVAL expr unit), ADDDATE(expr,days)

When invoked with the INTERVAL form of the second argument, ADDDATE() is a synonym for DATE_ADD(). The related function SUBDATE() is a synonym for DATE_SUB(). For information on the INTERVAL unit argument, see the discussion for DATE_ADD().
```

so that u also need to support `ADDDATE(expr,days)`, but it is very easy:)

**Comment by zimulala** [comment]
Gotcha.

**Comment by zimulala** [comment]
PTAL @siddontang  @qiuyesuifeng

File: expression/date_arith.go [hunk 1]
```diff
@@ -126,6 +146,13 @@ func (da *DateArith) evalArgs(ctx context.Context, args map[interface{}]interfac
 	if iVal == nil || err != nil {
 		return mysql.ZeroTimestamp, 0, 0, 0, 0, errors.Trace(err)
 	}
+	// handle adddate(expr,days) or subdate(expr,days) form
+	if da.Form == DateArithDaysForm {
```
Comment (qiuyesuifeng) [inline expression/date_arith.go]: Seems u have already get iVal in line 146,  how about define iVal first, then use `da.Form == DateArithDaysForm` condition?

File: expression/date_arith.go [hunk 2]
```diff
@@ -137,3 +164,19 @@ func (da *DateArith) evalArgs(ctx context.Context, args map[interface{}]interfac
...
+func (da *DateArith) evalDaysForm(val interface{}) (interface{}, error) {
+	switch val.(type) {
+	case string:
+		if strings.ToLower(val.(string)) == "false" {
+			return 0, nil
+		}
+		if strings.ToLower(val.(string)) == "true" {
+			return 1, nil
+		}
+		reg := regexp.MustCompile(`[\d]+`)
```
Comment (qiuyesuifeng) [inline expression/date_arith.go]: Can types.ToInt64 function hold regexp check?

File: expression/date_arith.go (hunk 1 above)
Comment (zimulala) [inline expression/date_arith.go]: handle adddate or subdate function with days form.

File: ast/functions.go [hunk 3]
```diff
@@ -338,26 +338,40 @@ func (n *FuncTrimExpr) IsStatic() bool {
...
 type DateArithType byte
 
 const (
+	// AddDate is to run adddate function option.
+	// See: https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_adddate
+	AddDate DateArithType = iota + 1
 	// DateAdd is to run date_add function option.
 	// See: https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_date-add
-	DateAdd DateArithType = iota + 1
+	DateAdd
```
Comment (siddontang) [inline ast/functions.go]: AddDate and DateAdd seem confused. use only one. we can use data_add and date_sub inner, no need to case other synonyms.

File: expression/date_arith.go [hunk 4]
```diff
@@ -28,24 +29,33 @@ import (
 type DateArithType byte
 
 const (
+	// AddDate is to run adddate function option.
+	// See: https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_adddate
+	AddDate DateArithType = iota + 1
```
Comment (siddontang) [inline expression/date_arith.go]: why define the duplicated Type here? can we combine them?

File: expression/date_arith.go [hunk 5]
```diff
@@ -28,24 +29,33 @@ import (
...
+	SubDate
+	// DateArithDaysForm is to run adddate or subdate function with days form Flag.
+	DateArithDaysForm
 )
 
 // DateArith is used for dealing with addition and substraction of time.
 type DateArith struct {
 	Op       DateArithType
-	Unit     string
+	Form     DateArithType
```
Comment (siddontang) [inline expression/date_arith.go]: what's the difference between Op and Form, add some comments for explanation.

File: ast/functions.go (hunk 3 above)
Comment (zimulala) [inline ast/functions.go]: but, here we has a String() interface, so I add this constant.

Comment (siddontang) [inline ast/functions.go]: use one is enough.

File: expression/date_arith.go [hunk 6]
```diff
@@ -137,3 +164,19 @@ func (da *DateArith) evalArgs(ctx context.Context, args map[interface{}]interfac
 
 	return t, years, months, days, durations, nil
 }
+
+func (da *DateArith) evalDaysForm(val interface{}) (interface{}, error) {
+	switch val.(type) {
```
Comment (siddontang) [inline expression/date_arith.go]: switch v := val.(type)

so you can use v directly later, no need to write val.(string).

File: expression/date_arith.go (hunk 2 above)
Comment (siddontang) [inline expression/date_arith.go]: define a global reg variable outer, no need to compile it each time.

File: expression/date_arith.go [hunk 7]
```diff
@@ -137,3 +164,19 @@ func (da *DateArith) evalArgs(ctx context.Context, args map[interface{}]interfac
 
 	return t, years, months, days, durations, nil
```
Comment (siddontang) [inline expression/date_arith.go]: too many return args, can we use a struct to enclose them?

Comment (zimulala) [inline expression/date_arith.go]: Gotcha.

File: expression/date_arith.go (hunk 2 above)
Comment (siddontang) [inline expression/date_arith.go]: how about if val is not number string? no error?

File: ast/functions.go (hunk 3 above)
Comment (zimulala) [inline ast/functions.go]: if so String() doesn't know adddate or dateadd operation.

Comment (siddontang) [inline ast/functions.go]: no need to distinguish them, they are the same.

File: expression/date_arith.go (hunk 4 above)
Comment (zimulala) [inline expression/date_arith.go]: It will be done by coocood later.

File: ast/functions.go (hunk 3 above)
Comment (zimulala) [inline ast/functions.go]: got it.

File: expression/date_arith.go (hunk 5 above)
Comment (zimulala) [inline expression/date_arith.go]: if remove `AddDate` and `SubDate` I think the `Form`  also can be removed.

File: expression/date_arith.go (hunk 2 above)
Comment (zimulala) [inline expression/date_arith.go]: yes, a global reg is better.
`types.ToInt64(val)` will have error.

**Comment by qiuyesuifeng** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Pull request merged**